src/controller/CommissioningWindowParams.h

```diff
@@ -0,0 +1,185 @@
+    System::Clock::Seconds16 mTimeout = System::Clock::Seconds16(300);
+    uint32_t mIteration               = 1000;
+    uint16_t mDiscriminator           = 0;
+};
+
+class CommissioningWindowPasscodeParams : public CommissioningWindowCommonParams<CommissioningWindowPasscodeParams>
+{
+public:
+    CommissioningWindowPasscodeParams() = default;
+
+    bool HasSetupPIN() const { return mSetupPIN.HasValue(); }
+    // Get the value of salt if present.
+    // Returns 0 if absent, make sure to check HasSetupPIN() if a valid value is required.
+    uint32_t GetSetupPIN() const { return mSetupPIN.ValueOr(0); }
+    // The setup PIN (Passcode) to use. A random one will be generated if not provided.
+    CommissioningWindowPasscodeParams & SetSetupPIN(uint32_t setupPIN) { return SetSetupPIN(MakeOptional(setupPIN)); }
+    // The setup PIN (Passcode) to use. A random one will be generated if NullOptional is used.
+    CommissioningWindowPasscodeParams & SetSetupPIN(Optional<uint32_t> setupPIN)
+    {
+        mSetupPIN = setupPIN;
+        return *this;
+    }
+
+    bool HasSalt() const { return mSalt.HasValue(); }
+    // Get the value of salt if present.
+    // Dies if absent! Make sure to check HasSalt()
+    ByteSpan GetSalt() const { return mSalt.Value(); }
+    // The salt to use. A random one will be generated if not provided.
+    // If provided, must be at least kSpake2p_Min_PBKDF_Salt_Length bytes
+    // and at most kSpake2p_Max_PBKDF_Salt_Length bytes in length.
+    CommissioningWindowPasscodeParams & SetSalt(ByteSpan salt) { return SetSalt(MakeOptional(salt)); }
+    // The salt to use. A random one will be generated if NullOptional is used.
+    // If provided, must be at least kSpake2p_Min_PBKDF_Salt_Length bytes
+    // and at most kSpake2p_Max_PBKDF_Salt_Length bytes in length.
+    // Note that this an overloaded optional arg function to support existing APIs.
+    CommissioningWindowPasscodeParams & SetSalt(Optional<ByteSpan> salt)
+    {
+        mSalt = salt;
+        return *this;
+    }
+
+    bool GetReadVIDPIDAttributes() const { return mReadVIDPIDAttributes; }
+    // Should the API internally read VID and PID from the device while opening the
+    // commissioning window.  If this argument is `true`, the API will read VID and PID
+    // from the device and include them in the setup payload passed to the callback.
+    CommissioningWindowPasscodeParams & SetReadVIDPIDAttributes(bool readVIDPIDAttributes)
+    {
+        mReadVIDPIDAttributes = readVIDPIDAttributes;
+        return *this;
+    }
+
+    Callback::Callback<OnOpenCommissioningWindow> * GetCallback() const { return mCallback; }
+    // The function to be called on success or failure of opening the commissioning window.
+    // This will include the SetupPayload generated from provided parameters.
+    CommissioningWindowPasscodeParams & SetCallback(Callback::Callback<OnOpenCommissioningWindow> * callback)
+    {
+        mCallback = callback;
```

Comment:
See, so here: we need to ensure that this got set, in the thing that uses these params, right?

Reply:
Params class can be re-used without callback from the consumer so `nullptr` can be a valid default here. 

Also, this has been existing behavior on the current API if the caller isn't interested in knowing the result for any reason - a nullptr callback can be supplied which isn't validated during request acceptance and ofcourse doesn't get notified by `OnOpenCommissioningWindowSuccess/OnOpenCommissioningWindowFailure` when it's done.

Reply:
Ah, OK.  If we support no callback, great.